proposals/0178-static-syscalls.md

```diff
@@ -0,0 +1,146 @@
+type: Core
+status: Review
+created: 2024-09-27
+---
+
+## Summary
+
+This SIMD introduces a new instruction syscall in the SBPF instruction set to 
+represent syscalls. Such a change aims to remove relocations when resolving 
+syscalls and simplify the instruction set, allowing for the straightforward 
+differentiation between external and internal calls. In addition, it proposes 
+a new `return` instruction to supersede the `exit` instruction.
+
+## Motivation
+
+The resolution of syscalls during ELF loading requires relocating addresses, 
+which is a performance burden for the validator. Relocations require an entire 
+copy of the ELF file in memory to either relocate addresses we fetch from the 
+symbol table or offset addresses to after the start of the virtual machine’s 
+memory. Moreover, relocations pose security concerns, as they allow the 
+arbitrary modification of program headers and programs sections. A new 
+separate opcode for syscalls modifies the behavior of the ELF loader, allowing 
+us to resolve syscalls without relocations.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+The following must go into effect if and only if a program indicates the SBPF 
+version XX or higher in its ELF header e_flags field, according to the 
+specification of SIMD-0161.
+
+### New syscall instruction
+
+We introduce a new instruction in the SBPF instruction set, which we call 
+`syscall`. It must be associated with all syscalls in the SBPF format. Its 
+encoding consists of an opcode `0x95` and an immediate, which must refer to a 
+previously registered syscall. For more reference on the SBF ISA format, see 
+the 
+[spec document](https://github.com/solana-labs/rbpf/blob/main/doc/bytecode.md).
+
+For simplicity, syscalls must be represented as a natural number greater than 
+zero, so that they can be organized in a lookup table. This choice allows for 
+quick retrieval of syscall information from integer indexes. An instruction 
+`syscall 2` must represent a call to the function registered at position two 
+in the lookup table.
+
+Consequently, system calls in the Solana SDK and in any related compiler tools 
+must be registered as function pointers, whose address is a natural number 
+greater than zero, representing their position in a syscall lookup table. The 
+verifier must enforce that the immediate of a syscall instruction points to a 
+valid syscall, and throw `VerifierError::InvalidSyscall` otherwise.
+
+This new instruction comes together with modifications in the verification 
+phase. `call imm` (opcode `0x85`) instructions must only refer to internal 
+calls and its immediate field must only be interpreted as a relative address 
+to jump from the program counter. 
+
+### New return instruction
+
+The opcode `0x9D` must represent the return instruction, which supersedes the 
+`exit` instruction. The opcode (opcode `0x95`), previously assigned to the 
+`exit` instruction, must now be interpreted as the new syscall instruction.
```

Comment:
Also, changing the name from `exit` to `return` when it is the same instruction could be confusing (I have already seen this confused in other SIMDs).

Reply:
Side note - we should bundle large sets of proposed ISA changes together into the same SBPF version upgrade, so that clients don't have to support a mis-mash of ISAs based on feature flags.

Reply:
Motivation is that `exit` was occupying the slot in the instruction class for controlflow with immediate values and it does not take an immediate value. The new `syscall` opcode however does, so it took its place.